### PR TITLE
:has(:lang(~)) doesn't get invalidated

### PR DESCRIPTION
**Pull request by rniwa**
#### 0c6665bbc95114525dc42754787d5e2527a51b94
<pre>
:has(:lang(~)) doesn&apos;t get invalidated
<a href="https://bugs.webkit.org/show_bug.cgi?id=243172">https://bugs.webkit.org/show_bug.cgi?id=243172</a>

Reviewed by Antti Koivistso.

Make container query work with :lang pseudo class.

This patch introduces ElementRareData::effectiveLang to store the language of each element,
and uses Style::PseudoClassChangeInvalidation to invalidate styles for :lang pseudo class.

As an optimization, this patch also introduces Document::effectiveDocumentElementLanguage
which is the effective language of the document element as many websites specify a language there.
This avoids creating ElementRareData on all elements in the vast majority of cases.

* LayoutTests/fast/css/lang-pseudo-container-query-document-element-invalidation-expected.html: Added.
* LayoutTests/fast/css/lang-pseudo-container-query-document-element-invalidation.html: Added.pee
* LayoutTests/fast/css/lang-pseudo-container-query-invalidation-expected.html: Added.
* LayoutTests/fast/css/lang-pseudo-container-query-invalidation-xhtml-expected.html: Added.
* LayoutTests/fast/css/lang-pseudo-container-query-invalidation-xhtml.xhtml: Added.
* LayoutTests/fast/css/lang-pseudo-container-query-invalidation.html: Added.
* LayoutTests/fast/css/lang-pseudo-container-query-multiple-document-elements-invalidation-expected.html: Added.
* LayoutTests/fast/css/lang-pseudo-container-query-multiple-document-elements-invalidation.html: Added.

* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesLangPseudoClass): Use newly introduced effectiveLang.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::childrenChanged):
(WebCore::Document::effectiveDocumentElementLanguage): Added.
(WebCore::Document::setDocumentElementLanguage): Added.
* Source/WebCore/dom/Document.h:

* Source/WebCore/dom/Element.cpp:
(WebCore::effectiveLangFromAttribute):
(WebCore::Element::attributeChanged): Added the style invalidation logic for :lang.
(WebCore::Element::insertedIntoAncestor): Propagate the lang attribute from the parent.
(WebCore::Element::removedFromAncestor): Clear the effective lang when appropriate.
(WebCore::Element::computeInheritedLanguage const): Deleted.
(WebCore::Element::effectiveLang const): Added.
(WebCore::Element::langFromAttribute const): Added.
(WebCore::Element::locale const):

* Source/WebCore/dom/Element.h:

* Source/WebCore/dom/ElementRareData.cpp:
(WebCore::SameSizeAsElementRareData):

* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::effectiveLang const): Added.
(WebCore::ElementRareData::setEffectiveLang): Added.
(WebCore::ElementRareData::useTypes const):

* Source/WebCore/dom/NodeRareData.h:
* Source/WebCore/style/PseudoClassChangeInvalidation.cpp:
(WebCore::Style::PseudoClassChangeInvalidation::computeInvalidation):
(WebCore::Style::PseudoClassChangeInvalidation::collectRuleSets):
* Source/WebCore/style/PseudoClassChangeInvalidation.h:
(WebCore::Style::PseudoClassChangeInvalidation::Value): Added.
(WebCore::Style::PseudoClassChangeInvalidation::PseudoClassChangeInvalidation): Added a new variant
which takes AnyValueTag as an argument.

* Source/WebCore/html/BaseDateAndTimeInputType.cpp:
(WebCore::BaseDateAndTimeInputType::localeIdentifier const):
(WebCore::BaseDateAndTimeInputType::setupDateTimeChooserParameters):

* Tools/TestWebKitAPI/Tests/WebCore/DocumentOrder.cpp:
(TestWebKitAPI::createDocument): Initialize XMLNames via ProcessWarming.

Canonical link: <a href="https://commits.webkit.org/253764@main">https://commits.webkit.org/253764@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/48d597b0403cc4cd896f8fa43ae158534afc960b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86960 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31046 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17821 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95827 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149540 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/90938 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29410 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79098 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91004 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92576 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23746 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73801 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23760 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78735 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79029 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66766 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/78861 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27149 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/12870 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72523 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27087 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13884 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25858 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2661 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28768 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36734 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75311 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28710 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33161 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16662 "Passed tests") |

### CODE REVIEW

File: Tools/TestWebKitAPI/Tests/WebCore/DocumentOrder.cpp [hunk 1]
```diff
@@ -52,9 +53,7 @@ using namespace WebCore;
 
 static Ref<Document> createDocument()
 {
-    HTMLNames::init();
-    WebKitFontFamilyNames::init();
-    initializeCommonAtomStrings();
+    ProcessWarming::initializeNames();
```
Comment (anttijk) [inline Tools/TestWebKitAPI/Tests/WebCore/DocumentOrder.cpp]: Unrelated change?

File: Source/WebCore/dom/Document.h [hunk 2]
```diff
@@ -1928,6 +1931,7 @@ class Document
     bool m_hasXMLDeclaration { false };
 
     AtomString m_contentLanguage;
+    AtomString m_effectiveDocumentElementLanguage; // DocumentElement's lang attribute value.
```
Comment (anttijk) [inline Source/WebCore/dom/Document.h]: The variable name could be just m_documentElementLanguage as it is the actual value rather than an effective value. The accessor however returns an effective value (and is correctly named) because it falls back to m_contentLanguage.

**Review by anttijk — APPROVED** [review]

File: Source/WebCore/dom/Document.cpp [hunk 3]
```diff
@@ -1507,6 +1508,23 @@ void Document::setContentLanguage(const AtomString& language)
...
+
+void Document::setEffectiveDocumentElementLanguage(const AtomString& language)
+{
+    if (m_effectiveDocumentElementLanguage == language)
+        return;
+    m_effectiveDocumentElementLanguage = language;
+
+    // Recalculate style so language is used when selecting the initial font.
+    m_styleScope->didChangeStyleSheetEnvironment();
+}
```
Comment (anttijk) [inline Source/WebCore/dom/Document.cpp]: We can also skip style update if m_effectiveDocumentElementLanguage.isNull() &&  m_contentLanguage ==  language (that is if effectiveDocumentElementLanguage() doesn't change)

**Review by anttijk — APPROVED** [review]

File: Source/WebCore/dom/Document.h (hunk 2 above)
Comment (rniwa) [reply Source/WebCore/dom/Document.h]: I guess setter should also be setDocumentElementLanguage.
